wit/types.wit

```diff
@@ -0,0 +1,158 @@
+      timeout-error(string),
+      protocol-error(string),
+      status-error(u16),
+      unexpected-error(string)
+  }
+
+  // This following block defines the `fields` resource which corresponds to
+  // HTTP standard Fields. Soon, when resource types are added, the `type
+  // fields = u32` type alias can be replaced by a proper `resource fields`
+  // definition containing all the functions using the method syntactic sugar.
+  type fields = u32
+  drop-fields: func(fields: fields)
+  new-fields: func(entries: list<tuple<string,string>>) -> fields
+  fields-get: func(fields: fields, name: string) -> list<string>
+  fields-set: func(fields: fields, name: string, value: list<string>)
+  fields-delete: func(fields: fields, name: string)
+  fields-append: func(fields: fields, name: string, value: string)
+  fields-entries: func(fields: fields) -> list<tuple<string,string>>
+  fields-clone: func(fields: fields) -> fields
+
+  type headers = fields
+  type trailers = fields
+
+  // The following block defines the `body` type which corresponds to the HTTP
+  // standard Contents. With Preview3, all of these fields can be replaced by a
+  // single type definition:
+  //
+  //   type body = stream<u8, option<trailers>>
+  //
+  // In the interim, we need to use separate `input-stream` and `output-stream`
+  // resource types defined by `wasi:io/streams`. The `finish-` functions
+  // emulate the stream's result value and MUST be called exactly once after
+  // the final read/write from/to the stream before dropping the stream.
+  type incoming-body = input-stream
+  type outgoing-body = output-stream
+  finish-incoming-body: func(body: incoming-body) -> option<trailers>
+  finish-outgoing-body: func(body: outgoing-body, trailers: option<trailers>)
+
+  // The following block defines the `incoming-request` and `outgoing-request`
+  // resource types that correspond to HTTP standard Requests. Soon, when
+  // resource types are added, the `u32` type aliases can be replaced by
+  // proper `resource` type definitions containing all the functions as
+  // methods. Later, Preview2 will allow both types to be merged together into
+  // a single `request` type (that uses the single `body` type mentioned
+  // above). The `consume` and `write-body` methods may only be called once
+  // (and return failure thereafter).
+  type incoming-request = u32
+  type outgoing-request = u32
+  drop-incoming-request: func(request: incoming-request)
+  drop-outgoing-request: func(request: outgoing-request)
+  incoming-request-method: func(request: incoming-request) -> method
+  incoming-request-path: func(request: incoming-request) -> string
+  incoming-request-scheme: func(request: incoming-request) -> option<scheme>
+  incoming-request-authority: func(request: incoming-request) -> string
+  incoming-request-headers: func(request: incoming-request) -> headers
+  incoming-request-consume: func(request: incoming-request) -> result<incoming-body>
```

Comment:
The body is optional, and it's absent in 90%++ of requests.

However, it seems that this API assumes it's always present (presumably, a missing body is represented as an empty stream with EOF, but I don't think that's a good choice), and it requires calls to `consume` and `finish` to get `trailers`, which adds a lot of unnecessary overhead.

Reply:
Ah, makes sense.  So would a solution be to have `consume` return a `result` containing a 3-case variant such as:
```
variant {
  body(incoming-stream),
  trailers(trailers),
  none
}
```
(and similarly for `write`, but outgoing)?

Reply:
Filed #6 to consider further.